src/controller/CHIPDeviceController.cpp

```diff
@@ -228,12 +234,68 @@ CHIP_ERROR DeviceController::Init(NodeId localDeviceId, ControllerInitParams par
     mState         = State::Initialized;
     mLocalDeviceId = localDeviceId;
 
+    VerifyOrExit(params.operationalCredentialsDelegate != nullptr, err = CHIP_ERROR_INVALID_ARGUMENT);
+    mOperationalCredentialsDelegate = params.operationalCredentialsDelegate;
+
+    err = LoadLocalCredentials(admin);
+    SuccessOrExit(err);
+
     ReleaseAllDevices();
 
 exit:
     return err;
 }
 
+CHIP_ERROR DeviceController::LoadLocalCredentials(Transport::AdminPairingInfo * admin)
+{
+    ChipLogProgress(Controller, "Getting operational keys");
+    Crypto::P256Keypair * keypair = admin->GetOperationalKey();
+
+    ReturnErrorCodeIf(keypair == nullptr, CHIP_ERROR_NO_MEMORY);
+
+    if (!admin->AreCredentialsAvailable())
+    {
+        chip::Platform::ScopedMemoryBuffer<uint8_t> buffer1;
+        ReturnErrorCodeIf(!buffer1.Alloc(kMaxCHIPOpCertLength), CHIP_ERROR_NO_MEMORY);
+
+        chip::Platform::ScopedMemoryBuffer<uint8_t> buffer2;
+        ReturnErrorCodeIf(!buffer2.Alloc(kMaxCHIPOpCertLength), CHIP_ERROR_NO_MEMORY);
+
+        uint8_t * CSR    = buffer1.Get();
+        size_t csrLength = kMaxCHIPOpCertLength;
+        ReturnErrorOnFailure(keypair->NewCertificateSigningRequest(CSR, csrLength));
+
+        uint8_t * cert   = buffer2.Get();
+        uint32_t certLen = 0;
+
+        ChipLogProgress(Controller, "Generating operational certificate for the controller");
+        ReturnErrorOnFailure(mOperationalCredentialsDelegate->GenerateNodeOperationalCertificate(
+            PeerId().SetNodeId(mLocalDeviceId), ByteSpan(CSR, csrLength), 1, cert, kMaxCHIPOpCertLength, certLen));
+
+        uint8_t * chipCert   = buffer1.Get();
+        uint32_t chipCertLen = 0;
+        ReturnErrorOnFailure(ConvertX509CertToChipCert(cert, certLen, chipCert, kMaxCHIPOpCertLength, chipCertLen));
```

Comment:
1. FabricID and NodeID should be updated based on incoming operational cert. That's how they are set.

2. Need to add a TODO about validating that the cert that came from the previous step, because:

  - It has to match the public key of the operational keypair from last CSR
  - It has to chain back to the trusted root
  
This is also true for the commissioning code on the client.

Right now I guess it's "maintained" by assumptions of GenerateNodeOperationalCertificate, but if that code does funky stuff at some point, very much good luck in trying to figure-out why a CASE session fails to establish.

Reply:
Added the TODO.